packages/shared/src/lib/group_utils.ts

```diff
@@ -0,0 +1,71 @@
+// Perhaps we can remove some of these functions if it becomes cumbersome to
+// implement these functions every time.  For example, `forEach` can be replaced
+// by map if we ignore the output
+interface Fillable<K, V> {
+  at(index: K): V | undefined;
+  set(index: K, val: V): void;
+  map<V2>(f: (val: V) => V2): Fillable<K, V2>;
+  neighbors(index: K): K[];
+  isInBounds(index: K): boolean;
+  forEach(f: (value: V, index: K) => void): void;
+}
+
+export function getGroup<K, V>(index: K, g: Fillable<K, V>): K[] {
+  if (!g.isInBounds(index)) {
+    return [];
+  }
+
+  const ret: K[] = [];
+
+  // The type assertion (as V) is regrettable, but I can't think
+  // of a better way to satisfy the strict null checks
+  // At any rate, calling floodDo on an out-of-bounds index should be a no-op
+  const starting_value = g.at(index) as V;
```

Comment:
If `at()` returned `undefined` for out-of-bounds indices, we could use something like this:
```
const startingValue = g.at(index);
if (startingValue === undefined) {
  return [];
}
```

Reply:
Related to my other comment, I would expect to get a non-empty group for in-bounds `undefined`:

```
const my_grid = Grid.from2DArray([
    [undefined, 0, 1],
    [undefined, 0, 1],
    [undefined, 0, 1],
])
getGroup({ x: 0, y: 0 }, my_grid).length  // 3
```

Reply:
Okay, so we'd have to prevent `undefined` to be valid for `V`, I didn't think about that before. There's always `null` to use, but just to get rid of `as V` isn't a great reason to do so.

Reply:
Yep I think that sounds right, and thanks for thinking on solutions for this too.  The only other "solution" I've thought about was to change the typing to say `.at()` returns `T` instead of `T | undefined`.  It would, of course, be a lie, but [it is one of the lies TypeScript makes for array accesses in the name of dev experience](https://stackoverflow.com/questions/50647399/typescript-accessing-an-array-element-does-not-account-for-the-possibility-of-u) 😆